### ISSUE TITLE
Exploration: use idle mesh-node compute for bounded LoRA fine-tuning jobs

### ISSUE THREAD

**Issue by michaelneale** [Mesh-LLM/mesh-llm#1923]
> [!NOTE]
> This is an **exploration proposal**, not a roadmap commitment or a claim that the current fleet model can already be tuned. The goal is to learn whether Mesh can safely turn otherwise-idle node compute into useful small-model post-training jobs.

## Idea

Allow an opted-in, capable Mesh node—initially a single Apple Silicon node such as a 128 GB M5-class machine—to use idle accelerator time for a bounded LoRA/QLoRA job. The node would train a small specialist from an existing instruct checkpoint, evaluate it against the unchanged base, and emit a reproducible adapter bundle plus an optional GGUF that Mesh can serve.

This should be **single-node training coordinated by Mesh**, not distributed gradient training. The mesh-specific value is that Mesh can supply teacher models and verified trajectories, select an eligible idle node, track the long-running job, preserve artifact identity, and serve a promoted result.

The working v1 hypothesis is a dense **1B–9B model**, not full fine-tuning and not the current 35B Qwen model.

## What could a specialist learn?

Candidate recipes, in likely value order:

1. **Verified tool-use / format controller** — train on successful or corrected tool trajectories, including tool schemas and deterministic task verdicts. This could improve tool selection, argument shape, retry/recovery behavior, structured output, and stopping.
2. **Distillation from stronger mesh models** — use a large mesh model to produce candidate answers or trajectories, retain only verifier-approved examples, and train a smaller, cheaper specialist.
3. **House style or response policy** — teach stable tone, layout, terminology, classification, extraction, or exact output formats from an explicitly approved dataset.
4. **Project procedure specialist** — generate source-grounded instruction pairs from approved docs and runbooks. Tune durable procedures and vocabulary; keep changing facts and citations in retrieval.
5. **Preference tuning later** — only after SFT exposes repeatable chosen/rejected errors and a suitable trainer/backend is proven.

The clearest first product experiment is a verified tool/format controller because it has a deterministic success metric. A persona/style adapter is the easiest smoke demo. “Train on this project” is attractive, but should not be presented as reliable storage for mutable knowledge.

## Proposed user and system shape

A fine-tune is a persistent job, not an inference request:

```text
mesh-llm tune plan --base <model-ref> --recipe sft-lora --data <path>
mesh-llm tune run <plan-or-spec>
mesh-llm tune status <job-id>
mesh-llm tune cancel <job-id>
mesh-llm tune promote <job-id>
```

Suggested durable states:

```text
queued -> preflight -> waiting_for_idle -> materializing -> training
       -> evaluating -> packaging -> ready_for_promotion -> promoted
```

`cancelled`, `suspended`, and `failed` are explicit alternatives. Each transition should survive a host/plugin restart.

“Idle” needs a real admission policy, not a best-effort background process:

- owner opt-in and an explicit node/resource budget;
- no active inference using the accelerator;
- reserved memory and disk headroom;
- maximum wall time, temperature/power policy, and checkpoint interval;
- inference remains higher priority: checkpoint and suspend/cancel according to policy when the node is needed;
- do not silently pause serving, download private data, or schedule paid remote compute;
- avoid co-resident Metal inference during training, especially while the current upstream stability issue remains open.

## Architecture sketch

The first implementation should fit the existing plugin architecture as a stable capability such as `model-trainer.v1`:

- the trainer plugin owns backend-specific training logic, local job state, logs, checkpoints, and evaluation runners;
- the Mesh host owns plugin lifecycle, permissions, resource admission, API/CLI projection, artifact identity, and promotion;
- control operations remain small; datasets, checkpoints, and model artifacts use paths or negotiated side streams rather than the plugin control connection;
- MLX-LM is the first Apple backend, not part of the public contract;
- the plugin exposes typed recipes, never an arbitrary remote shell command.

Start locally with no mesh protocol change. If the local flow proves useful, nodes could later gossip an **additive, optional** training capability: backend, supported model families/recipes, allocatable memory, free disk, and busy/idle status. Remote execution must remain owner-approved and signed. Datasets must not be put into ordinary gossip.

## Data and privacy contract

Inputs should be explicit, consented JSONL (`chat`, `tools`, `completions`, or `text`) with train/validation/test hashes and a separate provenance manifest.

Do not automatically scrape channels, DMs, repositories, or local files. Source selection, participant consent, secret/PII filtering, licensing, retention, and deletion are product requirements. Split whole tasks/trajectories—not individual turns—to avoid evaluation leakage.

## Model and artifact identity

A served GGUF is not the differentiable checkpoint used by MLX-LM. A tune specification must distinguish:

1. `base_model_identity` — logical architecture, tokenizer, revision, and hashes;
2. `training_checkpoint` — the compatible Hugging Face/MLX checkpoint;
3. `serving_artifact` — the GGUF or package Mesh will load.

The job must prove those identities correspond; matching display names are not enough.

The canonical result should be an immutable adapter bundle containing:

- adapter weights/config;
- base, tokenizer, trainer, and data identities;
- seed and hyperparameters;
- metrics and checkpoint-selection reason;
- base-versus-tuned product evaluation through the real prompt/template path;
- licenses and provenance declarations;
- optional fused checkpoint;
- optional promoted GGUF/package, checksum, and quantization recipe.

Loss alone is not a promotion gate. Promotion should be a separate explicit action after the tuned model beats the base on a held-out task metric without material regressions.

## What current code and tools make plausible

Verified against Mesh `main` at [`7c69459`](https://github.com/Mesh-LLM/mesh-llm/commit/7c69459d15959ed85e74644b06eb165599ebb160):

- Mesh's [plugin architecture](https://github.com/Mesh-LLM/mesh-llm/blob/7c69459d15959ed85e74644b06eb165599ebb160/docs/plugins/README.md) already separates lifecycle/control from large side-stream transfers and supports stable capability contracts.
- `model-artifact` already distinguishes [portable model identity from local artifact paths](https://github.com/Mesh-LLM/mesh-llm/blob/7c69459d15959ed85e74644b06eb165599ebb160/crates/model-artifact/README.md).
- The native SafeTensors→GGUF metadata path supports dense Qwen2/Qwen3, while explicitly rejecting Qwen3.5/Qwen3Next/Qwen3VL pending architecture-specific support ([source](https://github.com/Mesh-LLM/mesh-llm/blob/7c69459d15959ed85e74644b06eb165599ebb160/crates/skippy-model/src/gguf_template.rs#L86-L132)).
- `hardware.lora_adapters` exists in configuration, but current wiring status marks it `Unwired` because there is [no native LoRA load/apply path](https://github.com/Mesh-LLM/mesh-llm/blob/7c69459d15959ed85e74644b06eb165599ebb160/crates/mesh-llm-config/src/wiring_status.rs#L842-L848). Direct adapter serving is therefore not a v1 assumption.

Current MLX-LM can train LoRA/QLoRA and save/fuse a checkpoint ([guide](https://github.com/ml-explore/mlx-lm/blob/9d1e356e7cc6549e7d1697adabe2ea01ff8e062c/mlx_lm/LORA.md)). Its built-in GGUF exporter is still limited to Llama/Mistral/Mixtral, even though fusion and checkpoint saving happen before that check ([source](https://github.com/ml-explore/mlx-lm/blob/9d1e356e7cc6549e7d1697adabe2ea01ff8e062c/mlx_lm/fuse.py#L75-L108)).

That suggests two initial artifact paths:

- **Llama/Mistral:** MLX-LM LoRA -> fuse/dequantize -> built-in GGUF export -> Mesh serve.
- **Dense Qwen2/Qwen3:** MLX-LM LoRA -> fuse/dequantize to SafeTensors -> `skippy-quantize` BF16 GGUF conversion -> quantize -> Mesh serve. The concrete unknown is whether the fused MLX snapshot is complete and HF-layout-compatible end to end.

An MLX adapter cannot currently be assumed compatible with llama.cpp's PEFT/GGUF adapter tooling; naming, scale/alpha, orientation, and all target tensors would need a verified translator. Keep this as a fallback, not the primary first path.

### Gate 0A — smallest portable end-to-end proof

Use Llama 3.2 1B or 3B with a few thousand approved, verifier-backed tool/format examples:

1. pin base/training revisions, tokenizer hash, MLX/MLX-LM versions, and seed;
2. train LoRA/QLoRA on one M5-class node;
3. record peak memory, trained tokens/sec, total time, energy/thermal observations, and checkpoint behavior;
4. retain the adapter, fuse/dequantize, export GGUF, and hash all outputs;
5. compare base and tuned models through the real Mesh front door;
6. promote only if the task metric improves and regression/safety checks pass.

### Gate 0B — dense Qwen bridge

Repeat on a dense Qwen3 model around 4B–8B using the in-tree SafeTensors→BF16 GGUF and quantization path. Verify tokenizer/config completeness, full tensor-name/shape coverage, output equivalence, loadability, and real Mesh inference before calling the family supported.

### Gate 0C — 35B-A3B research only

Do not make this a v1 target. MLX-LM issue [#1185](https://github.com/ml-explore/mlx-lm/issues/1185) remains open and reports a Metal descriptor/resource-limit failure after tens to hundreds of optimizer steps on Qwen3.5-family 27B/35B-A3B training; current Mesh native conversion also rejects Qwen3.5 model types. Only revisit after the small paths pass, and require a long, repeatable stability run plus a separately proven serving artifact path.

### Gate 1 — local persistent trainer job

Add typed plan/preflight/run/status/cancel/resume operations, idle-resource admission, durable state, and restart-safe checkpoints behind `model-trainer.v1`.

### Gate 2 — evaluated artifact promotion

Integrate the immutable bundle with Mesh model identity/storage/package handling. Promotion remains explicit and gated on real serving and base-versus-tuned evaluation.

### Gate 3 — mesh-native recipe

Use stronger mesh models to generate trajectories and a deterministic verifier to filter them, then train a cheap local specialist. This is the first step that makes the feature more than a local trainer wrapper.

## Exploration success criteria

- One supported small base completes on an M5-class node under a declared resource budget.
- Training yields to higher-priority inference according to an explicit, tested policy.
- Cancel/resume and host/plugin restart behavior preserve clear artifact state.
- Adapter, manifest, data hashes, and evaluation report are reproducible.
- A promoted GGUF loads through the real Mesh serve path and improves a held-out product metric over the base.
- No data is silently harvested, gossiped, or published.
- The experiment records a clear go/no-go recommendation for a developer-facing `mesh-llm tune` primitive.

## Explicit non-goals for this exploration

- full-model training or training from scratch;
- multi-node gradient/optimizer sharding;
- tuning an arbitrary GGUF in place;
- promising 35B-A3B support;
- silently using channel/DM history or remote nodes;
- automatically publishing artifacts;
- committing the feature to the roadmap before the gates pass.

## Open questions

- Should the first usable surface be the low-level `mesh-llm tune` developer primitive, or a project/channel command? Recommendation: prove the developer primitive first.
- What exactly qualifies as idle, and should inference arrival suspend, cancel, or wait for the next checkpoint?
- Is an adapter-only local result useful before direct adapter serving exists, or is GGUF promotion required for v1?
- Which initial metric and dataset best demonstrate value: tool-call success, structured-output validity, routing/classification, or style adherence?
- Where should private bundles live, and what owner approval is required to move them between nodes?

## Research credit

Initial option survey and integration design were developed jointly by Dario and codexy in Buzz, then checked against the repository and current upstream MLX-LM behavior.
